src/error.rs

```diff
@@ -0,0 +1,43 @@
+use errno;
+use errno::{set_errno, Errno};
+use std::cell::RefCell;
+use std::fmt::Display;
+
+use crate::memory::{release_vec, Buffer};
+
+thread_local! {
+    static LAST_ERROR: RefCell<Option<String>> = RefCell::new(None);
```

Comment:
Using errno-style thread local for a modern C-library is not the best approach.

I *think* what you are supposed to do in modern C is using an out parameter for error, like this:

```C
struct ErrorInfo {
    char* message; // statically allcoated message, which doesn't need free
}

// 0 means success, != error
int frobnicate(int input1, int input2, my_struct* output, ErrorInfo* error);
```

The call site would look like this

```C
struct ErrorInfo error; // reused across many calls;

struct MyStruct res;
if !frobnicate(92, 62, &res, &error) {
    fprintf(2, "error: %s", error.message);
    return;
}
```

Reply:
I am pretty sure I am not thinking up, and that this is indeed written down in some blog post, but, alas, googling "modern C error management" doesn't help :(

Reply:
Thanks for the review.

On this particular point, I didn't really choose to use errno, this is what cgo recommends (and this lib is designed to be consumed with go):

`Any C function (even void functions) may be called in a multiple assignment context to retrieve both the return value (if any) and the C errno variable as an error (use _ to skip the result value if the function returns void). For example: `
https://golang.org/cmd/cgo/

The use of thread_local and RefCell was something I copied from another lib in order to get a string error message (not just a number) and seems to make having one static value less dangerous. But yes... I could definitely envision a better solution.

In particular, you suggest using the return value instead of errno (0 success, other failure), and then returning any result and error as pointer arguments. It seems to make uglier function definitions on one hand. On the other, it does remove some "magic" code and make the actual information passing explicity (not "put this error in a box for later and hope someone picks it up").

Reply:
Oh, if cgo requires errno for convenient API on the Go side, than our hands I tied. It maybe makes scene to add a comment like "we use errno because cgo has special integration with it". 

For getting messages out of errors, I would try, if possible, to restrict them to a set of static strings, so that the caller doesn't have to worry about allocation/deallocation. The API then may look like this:

```
pub extern "c" fn get_last_error() -> *const u8 { // returns static C-string or null, by decoding errno value
}
```

Reply:
Interesting point about static strings.
I just declare a whole set of them as consts in one file and reference them?

I could even use the &error approach to return the string in the same call, even with errno (that would just change the meaning of the return value)